compiler/rustc_codegen_ssa/src/mir/rvalue.rs

```diff
@@ -200,31 +198,25 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         assert!(src.layout.is_sized());
         assert!(dst.layout.is_sized());
 
-        if let Some(val) = self.codegen_transmute_operand(bx, src, dst.layout) {
-            val.store(bx, dst);
-            return;
-        }
-
-        match src.val {
-            OperandValue::Ref(..) | OperandValue::ZeroSized => {
-                span_bug!(
-                    self.mir.span,
-                    "Operand path should have handled transmute \
-                    from {src:?} to place {dst:?}"
-                );
-            }
-            OperandValue::Immediate(..) | OperandValue::Pair(..) => {
-                // When we have immediate(s), the alignment of the source is irrelevant,
-                // so we can store them using the destination's alignment.
-                src.val.store(bx, dst.val.with_type(src.layout));
-            }
+        if src.layout.size == dst.layout.size {
+            // Since in this path we have a place anyway, we can store or copy to it,
+            // making sure we use the destination place's alignment even if the
+            // source would normally have a higher one.
+            src.val.store(bx, dst.val.with_type(src.layout));
+        } else if src.layout.is_uninhabited() {
+            bx.unreachable()
+        } else {
+            // Since this is known statically and the input could have existed
+            // without already having hit UB, might as well trap for it, even
+            // though it's UB so we *could* also unreachable it.
+            bx.abort();
         }
     }
 
     /// Attempts to transmute an `OperandValue` to another `OperandValue`.
     ///
     /// Returns `None` for cases that can't work in that framework, such as for
-    /// `Immediate`->`Ref` that needs an `alloc` to get the location.
+    /// `Immediate`->`Ref` that needs an `alloca` to get the location.
     pub(crate) fn codegen_transmute_operand(
```

Comment:
There's one remaining caller for this, `codegen_rvalue_operand`. It doesn't have a fallback when this returns `None`.
- Why does it even make sense to still return an `Option` then?
- Why can that codepath never hit SIMD transmutes?

Reply:
Ah, `rvalue_creates_operand` actually *decides* which rvalue expressions can remain out-of-memory.

So, there's a contract here that `codegen_transmute_operand` must return `Some` for everything where `rvalue_creates_operand` returns `true`. That would be good to document.

Also, it may be worth removing the `Option` return value and having the panic inside this function to make it more clear that there is no fallback path.